### PR TITLE
User Notifications WebSocket

### PR DESCRIPTION
**Pull request by jeffersondarcy**

### CODE REVIEW

File: common/notification/index.ts [hunk 1]
```diff
@@ -11,11 +11,12 @@ import { Pupil } from '../entity/Pupil';
 import { assert } from 'console';
 import { triggerHook } from './hook';
 import { USER_APP_DOMAIN } from '../util/environment';
+import { inAppChannel } from './channels/inapp';
 
 const logger = getLogger('Notification');
 
 // This is the main extension point of notifications: Implement the Channel interface, then add the channel here
-const channels = [mailjetChannel];
+const channels = [mailjetChannel, inAppChannel];
```
Comment (Jonasdoubleyou) [inline common/notification/index.ts:19]: comment out

File: common/notification/index.ts [hunk 2]
```diff
@@ -244,33 +245,38 @@ async function createConcreteNotification(
...
+        authToken: legacyUser.authToken ?? '',
         USER_APP_DOMAIN,
     };
 
     try {
-        const channel = channels.find((it) => it.canSend(notification));
-        if (!channel) {
+        const user = getUserForTypeORM(legacyUser);
+        const channelsToSendTo = channels.filter((it) => it.canSend(notification, user));
+        if (!channelsToSendTo || channelsToSendTo.length === 0) {
```
Comment (Jonasdoubleyou) [inline common/notification/index.ts:264]: Can channelsToSendTo be null?

File: graphql/concrete_notification/fields.ts [hunk 3]
```diff
@@ -36,6 +39,32 @@ export class ExtendedFieldsConcreteNotificationResolver {
...
+    @FieldResolver((returns) => String)
+    @Authorized(Role.OWNER, Role.ADMIN)
+    async body() {
+        return `Mock Body Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut`;
+    }
+
+    // @TODO: this should be a field of ConcreteNotification
+    @FieldResolver((returns) => Date)
+    @Authorized(Role.OWNER, Role.ADMIN)
+    async createdAt(@Root() concreteNotification: ConcreteNotification) {
```
Comment (Jonasdoubleyou) [inline graphql/concrete_notification/fields.ts:57]: Can be removed?

File: graphql/concrete_notification/dummy_data.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { ConcreteNotification, ConcreteNotificationState } from '../../common/entity/ConcreteNotification';
+
+const ids = Array.from({ length: 10 }, (_, i) => i + 1);
+const dummyMinutes = [0, 5, 10, 15, 20, 30, 60, 120, 150, 60 * 24, 60 * 48];
+
+// Dummy dates indexed by dummy ids
+export const getDummyCreatedAt = (id: number): Date => new Date(Date.now() - dummyMinutes[id] * 1000 * 60);
```
Comment (Jonasdoubleyou) [inline graphql/concrete_notification/dummy_data.ts:7]: Can be removed (use sentAt)?

File: graphql/concrete_notification/fields.ts [hunk 5]
```diff
@@ -36,6 +39,32 @@ export class ExtendedFieldsConcreteNotificationResolver {
...
+        return getDummyCreatedAt(concreteNotification.id);
+    }
+
+    // TODO new resolver for getting concrete notifications by notificationId
```
Comment (Jonasdoubleyou) [inline graphql/concrete_notification/fields.ts:61]: Remove TODO, this is done?

File: graphql/user/fields.ts [hunk 6]
```diff
@@ -73,6 +74,12 @@ export class UserFieldsResolver {
         return fakeContext.user.roles;
     }
 
+    @FieldResolver((returns) => [ConcreteNotification])
+    @Authorized(Role.OWNER, Role.ADMIN)
+    async concreteNotifications(@Root() user: User): Promise<ConcreteNotification[]> {
+        return getDummyConcreteNotifications(user.userID);
```
Comment (Jonasdoubleyou) [inline graphql/user/fields.ts:80]: Add TODO

File: graphql/concrete_notification/fields.ts [hunk 7]
```diff
@@ -36,6 +39,32 @@ export class ExtendedFieldsConcreteNotificationResolver {
...
 
+    @FieldResolver((returns) => String)
+    @Authorized(Role.OWNER, Role.ADMIN)
+    async headline(@Root() concreteNotification: ConcreteNotification) {
+        return `Mock Headline ${concreteNotification.id}`;
+    }
+
+    @FieldResolver((returns) => String)
+    @Authorized(Role.OWNER, Role.ADMIN)
+    async body() {
```
Comment (Jonasdoubleyou) [inline graphql/concrete_notification/fields.ts:50]: Add TODO

**Review by Jonasdoubleyou — COMMENTED** [review]
Looks good at the moment, some minor cleanups. Feel free to also bring in Lomy's change for the last seen notifications. Do you want to Squash merge and then branch off a new branch from master or commit merge and continue to use the user-notifications branch?

**Comment by timidak** [comment]
> Do you want to Squash merge and then branch off a new branch from master or commit merge and continue to use the user-notifications branch?

I'd say squash merging and branching off again from `master` should be totally fine, @Jonasdoubleyou.

**Pull request closed without merging**